<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>6. DEG • scR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="6. DEG">
<meta property="og:description" content="scR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.8.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1_Pre-processing.html">1. Pre-processing</a>
    </li>
    <li>
      <a href="../articles/2_integration.html">2. Samples integration</a>
    </li>
    <li>
      <a href="../articles/3_annotation.html">3. Dataset annotation</a>
    </li>
    <li>
      <a href="../articles/4_visualisation.html">4. Visualisation</a>
    </li>
    <li>
      <a href="../articles/5_interoperability.html">5. Interoperability</a>
    </li>
    <li>
      <a href="../articles/6_DEG.html">6. DEG</a>
    </li>
    <li>
      <a href="../articles/7_Atlas.html">7. The Principles of Atlas building [WIP]</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>6. DEG</h1>
                        <h4 data-toc-skip class="author">Francesco
Carbone</h4>
            
            <h4 data-toc-skip class="date">Initial commit 14-04-2023
(rebuilt mars 2024)</h4>
      
      
      <div class="hidden name"><code>6_DEG.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="deg-for-dummies-and-biologists">DEG for dummies … and biologists<a class="anchor" aria-label="anchor" href="#deg-for-dummies-and-biologists"></a>
</h2>
<p>As a rule of thumb, DE test is usually performed on the uncorrected
counts rather than on the batch-corrected ones.</p>
<p>It is not ideal to run DE analysis on batch corrected values because
the correction introduces a dependency between expression values, which
will cause the violation of the indipendence assumption of every
statistical tests (e.g. every measurements are independent from each
other).</p>
<p>After performing the DE test it should be possible to determine if
the results are biologically relevant or if the technical variability is
affecting the results by masking the important biological events.</p>
<p>In the cases where the batch is strong, it is possible to account for
it during DE testing (like it is usually done in Bulk RNA-Seq) by</p>
<ol style="list-style-type: decimal">
<li>Setting <code>test.use</code> to either <strong>LR</strong>,
<strong>negbinom</strong>, <strong>poisson</strong>, or
<strong>MAST</strong>, and specifying the covariate variables with
<code>latent.vars= c("covariate1","covariate2",..)</code>.</li>
<li>Using <strong>monocle3</strong> and providing the design to
<strong>fit_models</strong>:
<code>fit_models(cds_object, model_formula_str = "~gender + batch")</code>.</li>
<li>Correcting the batch with external packages (e.g. scVI or Scanorama)
and running DE testing on the corrected counts.</li>
<li>Psudobulk</li>
</ol>
<div class="section level3">
<h3 id="to-regress-or-not-to-regress">To regress or not to regress<a class="anchor" aria-label="anchor" href="#to-regress-or-not-to-regress"></a>
</h3>
<p><code>vars.to.regress</code> is used to specify the variable to
regress out during scaling, so that their effect does not contribute too
much to the PCA.</p>
<div class="alert alert-info">
<p><strong>NOTE:</strong> Regressing out genes, is only used for PCA
dimension reduction, and does not affect DE testing.</p>
</div>
On the other hand, <code>latent.vars</code> in
<code><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers()</a></code> is used to pass variable that needs to be
included into the differential expression model selected.
<div class="alert alert-info">
<p><strong>NOTE:</strong> Latent variable will affect DE testing.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="assay-selection">Assay selection<a class="anchor" aria-label="anchor" href="#assay-selection"></a>
</h2>
<p><img src="6_DEG/single_sample.png"></p>
</div>
<div class="section level2">
<h2 id="a-word-on-the-umap">A word on the uMAP<a class="anchor" aria-label="anchor" href="#a-word-on-the-umap"></a>
</h2>
<p>When reading a umap, it is important to understand that:</p>
<ol style="list-style-type: decimal">
<li>The size of the cluster does not mean much</li>
<li>The global position of the clusters is better preserved in the umap
projection, but the distances between clusters are not meaningful</li>
</ol>
<p>In addition, the DE genes are not the genes driving the clustering
and the cell displacement in the umap. It is possible that there might
be a degree of overlap and similarity, but that’s all.<br>
To get an understanding of why the umap looks like this, we need to
check the genes in each dimension.</p>
<p>In bulk RNA-Seq, this is done by looking at the genes that contribute
to the different PCs.</p>
<p><img src="6_DEG/bulk_pcs.png"></p>
<p>Using a dummy dataset (<code>pbmc_small</code>), something similar is
also possible in single-cell</p>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 80</span></span>
<span><span class="co">## Number of edges: 2352</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.5533</span></span>
<span><span class="co">## Number of communities: 2</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span>
<span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 80</span></span>
<span><span class="co">## Number of edges: 2352</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.4014</span></span>
<span><span class="co">## Number of communities: 2</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span>
<span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 80</span></span>
<span><span class="co">## Number of edges: 2352</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.3074</span></span>
<span><span class="co">## Number of communities: 3</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span>
<span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 80</span></span>
<span><span class="co">## Number of edges: 2352</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.2374</span></span>
<span><span class="co">## Number of communities: 4</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span>
<span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 80</span></span>
<span><span class="co">## Number of edges: 2352</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.2111</span></span>
<span><span class="co">## Number of communities: 5</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<p><img src="6_DEG_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<pre><code><span><span class="co">## umap_ 1 </span></span>
<span><span class="co">## Positive:  SDPR, PPBP, PF4, TREML1, CA2, MYL9, TUBB1, PGRMC1, RUFY1, PARVB </span></span>
<span><span class="co">## Negative:  GNLY, AKR1C3, IGLL5, HLA-DQA1, CD1C, HLA-DPB1, RP11-290F20.3, VDAC3, S100A9, S100A8 </span></span>
<span><span class="co">## umap_ 2 </span></span>
<span><span class="co">## Positive:  RP11-290F20.3, HLA-DPB1, S100A9, S100A8, HLA-DQA1, IGLL5, CD1C, PARVB, VDAC3, RUFY1 </span></span>
<span><span class="co">## Negative:  SDPR, PF4, TREML1, PPBP, PGRMC1, GNLY, AKR1C3, TUBB1, CA2, MYL9</span></span></code></pre>
<p>The negative genes in both projections are the genes driving the
clustering of cluster 3 apart from the rest of the cells. Are these
genes genes DE in cluster 3? Maybe yes, maybe no. That is not the
point.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Francesco Carbone.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
