<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Improved Dotplot — Clustered_DotPlot_nl • scR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Improved Dotplot — Clustered_DotPlot_nl"><meta property="og:description" content="This function is a modified version of the Clustered DotPlots function from the scCustomize package.
The changes made are:"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.8.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/1_Pre-processing.html">1. Pre-processing</a>
    </li>
    <li>
      <a href="../articles/2_integration.html">2. Samples integration</a>
    </li>
    <li>
      <a href="../articles/3_annotation.html">3. Dataset annotation</a>
    </li>
    <li>
      <a href="../articles/4_visualisation.html">4. Visualisation</a>
    </li>
    <li>
      <a href="../articles/5_interoperability.html">5. Interoperability</a>
    </li>
    <li>
      <a href="../articles/6_DEG.html">6. DEG</a>
    </li>
    <li>
      <a href="../articles/7_Atlas.html">7. The Principles of Atlas building [WIP]</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Improved Dotplot</h1>
    
    <div class="hidden name"><code>Clustered_DotPlot_nl.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function is a modified version of the Clustered DotPlots function from the scCustomize package.
The changes made are:</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Clustered_DotPlot_nl</span><span class="op">(</span></span>
<span>  <span class="va">seurat_object</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  scale_dot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  colors_use_exp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#0000FF"</span>, <span class="st">"grey"</span>, <span class="st">"#FF4C4C"</span><span class="op">)</span>,</span>
<span>  exp_color_min <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  exp_color_middle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exp_color_max <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  print_exp_quantiles <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  colors_use_idents <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  x_lab_rotate <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  kcut <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  border <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  row_km_repeats <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  column_km_repeats <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  row_label_size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  col_label_size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  raster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot_km_elbow <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  elbow_kmax <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  idents <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  show_parent_dend_line <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ggplot_default_colors <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">666</span>,</span>
<span>  col_dendr <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  row_dendr <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  figure_title <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>seurat_object</dt>
<dd><p>Seurat object name.</p></dd>


<dt>features</dt>
<dd><p>Features to plot.</p></dd>


<dt>colors_use_exp</dt>
<dd><p>Color palette to use for plotting expression scale.  Default is `viridis::plasma(n = 20, direction = -1)`.</p></dd>


<dt>exp_color_min</dt>
<dd><p>Minimum scaled average expression threshold (everything smaller will be set to this).
Default is -2.</p></dd>


<dt>exp_color_middle</dt>
<dd><p>What scaled expression value to use for the middle of the provided `colors_use_exp`.
By default will be set to value in middle of `exp_color_min` and `exp_color_max`.</p></dd>


<dt>exp_color_max</dt>
<dd><p>Minimum scaled average expression threshold (everything smaller will be set to this).
Default is 2.</p></dd>


<dt>print_exp_quantiles</dt>
<dd><p>Whether to print the quantiles of expression data in addition to plots.
Default is FALSE.  NOTE: These values will be altered by choices of `exp_color_min` and `exp_color_min`
if there are values below or above those cutoffs, respectively.</p></dd>


<dt>colors_use_idents</dt>
<dd><p>specify color palette to used for identity labels.  By default if
number of levels plotted is less than or equal to 36 it will use "polychrome" and if greater than 36
will use "varibow" with shuffle = TRUE both from `DiscretePalette_scCustomize`.</p></dd>


<dt>x_lab_rotate</dt>
<dd><p>How to rotate column labels.  By default set to `TRUE` which rotates labels 45 degrees.
If set `FALSE` rotation is set to 0 degrees.  Users can also supply custom angle for text rotation.</p></dd>


<dt>row_km_repeats</dt>
<dd><p>`r lifecycle::badge("deprecated")` soft-deprecated.  See `feature_km_repeats`</p></dd>


<dt>column_km_repeats</dt>
<dd><p>`r lifecycle::badge("deprecated")` soft-deprecated.  See `ident_km_repeats`</p></dd>


<dt>row_label_size</dt>
<dd><p>Size of the feature labels.  Provided to `row_names_gp` in Heatmap call.</p></dd>


<dt>raster</dt>
<dd><p>Logical, whether to render in raster format (faster plotting, smaller files).  Default is FALSE.</p></dd>


<dt>plot_km_elbow</dt>
<dd><p>Logical, whether or not to return the Sum Squared Error Elbow Plot for k-means clustering.
Estimating elbow of this plot is one way to determine "optimal" value for `k`.
Based on: <a href="https://stackoverflow.com/a/15376462/15568251" class="external-link">https://stackoverflow.com/a/15376462/15568251</a>.</p></dd>


<dt>elbow_kmax</dt>
<dd><p>The maximum value of k to use for `plot_km_elbow`.  Suggest setting larger value so the
true shape of plot can be observed.  Value must be 1 less than number of features provided.  If NULL parameter
will be set dependent on length of feature list up to `elbow_kmax = 20`.</p></dd>


<dt>assay</dt>
<dd><p>Name of assay to use, defaults to the active assay.</p></dd>


<dt>group.by</dt>
<dd><p>Group (color) cells in different ways (for example, orig.ident).</p></dd>


<dt>idents</dt>
<dd><p>Which classes to include in the plot (default is all).</p></dd>


<dt>show_parent_dend_line</dt>
<dd><p>Logical, Sets parameter of same name in `ComplexHeatmap::Heatmap()`.
From `ComplexHeatmap::Heatmap()`: When heatmap is split, whether to add a dashed line to mark parent
dendrogram and children dendrograms.  Default is TRUE.</p></dd>


<dt>ggplot_default_colors</dt>
<dd><p>logical.  If `colors_use = NULL`, Whether or not to return plot using
default ggplot2 "hue" palette instead of default "polychrome" or "varibow" palettes.</p></dd>


<dt>seed</dt>
<dd><p>Sets seed for reproducible plotting (ComplexHeatmap plot).</p></dd>


<dt>col_dendr</dt>
<dd><p>Logical, cluster the genes (default is FALSE).</p></dd>


<dt>row_dendr</dt>
<dd><p>Logical, cluster the clusters (default is FALSE).</p></dd>


<dt>figure_title</dt>
<dd><p>Figure title column</p></dd>


<dt>flip</dt>
<dd><p>logical, whether to flip the axes of final plot.  Default is FALSE; rows = features and
columns = idents.It coul dbe because of a lot of You should add some details (e.g. figures and code used).</p></dd>


<dt>k</dt>
<dd><p>Value to use for k-means clustering on features  Sets (km) parameter in `ComplexHeatmap::Heatmap()`.
From `ComplexHeatmap::Heatmap()`: Apply k-means clustering on rows. If the value is larger than 1, the
heatmap will be split by rows according to the k-means clustering. For each row slice, hierarchical
clustering is still applied with parameters above.</p></dd>


<dt>feature_km_repeats</dt>
<dd><p>Number of k-means runs to get a consensus k-means clustering for features.
Note if `feature_km_repeats` is set to value greater than one, the final number of groups might be
smaller than row_km, but this might mean the original row_km is not a good choice.  Default is 1000.</p></dd>


<dt>ident_km_repeats</dt>
<dd><p>Number of k-means runs to get a consensus k-means clustering. Similar to
`feature_km_repeats`.  Default is 1000.</p></dd>


<dt>color_seed</dt>
<dd><p>random seed for the "varibow" palette shuffle if `colors_use = NULL` and number of
groups plotted is greater than 36.  Default = 123.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A ComplexHeatmap or if plot_km_elbow = TRUE a list containing ggplot2 object and ComplexHeatmap.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>* Removed the colomn annotation bar
* Added the option of clustering the identities (rows)</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p><a href="https://divingintogeneticsandgenomics.rbind.io/post/clustered-dotplot-for-single-cell-rnaseq/" class="external-link">https://divingintogeneticsandgenomics.rbind.io/post/clustered-dotplot-for-single-cell-rnaseq/</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Ming Tang (Original Code), Sam Marsh (Wrap single function, added/modified functionality) Francesco Carbone (Final tweakings)</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SeuratObject</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: sp</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘SeuratObject’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     intersect</span>
<span class="r-in"><span><span class="fu">Clustered_DotPlot_nl</span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc_small</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3E"</span>, <span class="st">"CD8"</span>, <span class="st">"GZMB"</span>, <span class="st">"MS4A1"</span><span class="op">)</span>,kcut <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> scCustomize v2.1.2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> If you find the scCustomize useful please cite.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> See 'samuel-marsh.github.io/scCustomize/articles/FAQ.html' for citation info.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: grid</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ComplexHeatmap version 2.18.0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Github page: https://github.com/jokergoo/ComplexHeatmap</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Documentation: http://jokergoo.github.io/ComplexHeatmap-reference</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> If you use it in published research, please cite either one:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> - Gu, Z. Complex Heatmap Visualization. iMeta 2022.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> - Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     genomic data. Bioinformatics 2016.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The new InteractiveComplexHeatmap package can directly export static </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> complex heatmaps into an interactive Shiny app with zero effort. Have a try!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> This message can be suppressed by:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   suppressPackageStartupMessages(library(ComplexHeatmap))</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> circlize version 0.4.16</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> CRAN page: https://cran.r-project.org/package=circlize</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Github page: https://github.com/jokergoo/circlize</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Documentation: https://jokergoo.github.io/circlize_book/book/</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> If you use it in published research, please cite:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Gu, Z. circlize implements and enhances circular visualization</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   in R. Bioinformatics 2014.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> This message can be suppressed by:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   suppressPackageStartupMessages(library(circlize))</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘dplyr’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     filter, lag</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     intersect, setdiff, setequal, union</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following requested variables were not found: CD8</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Scaling data with a low number of groups may produce misleading results</span>
<span class="r-plt img"><img src="Clustered_DotPlot_nl-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Francesco Carbone.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

