<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference • scR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.8.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/1_Pre-processing.html">1. Pre-processing</a>
    </li>
    <li>
      <a href="../articles/2_integration.html">2. Samples integration</a>
    </li>
    <li>
      <a href="../articles/3_annotation.html">3. Dataset annotation</a>
    </li>
    <li>
      <a href="../articles/4_visualisation.html">4. Visualisation</a>
    </li>
    <li>
      <a href="../articles/5_interoperability.html">5. Interoperability</a>
    </li>
    <li>
      <a href="../articles/6_DEG.html">6. DEG</a>
    </li>
    <li>
      <a href="../articles/7_Atlas.html">7. The Principles of Atlas building [WIP]</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="seurat-functions-scrna-seq-">Seurat functions (scRNA-Seq) <a href="#seurat-functions-scrna-seq-" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions using the seurat package</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="estimate_ambient_rna.html">estimate_ambient_rna()</a></code> </p>
        </td>
        <td><p>Estimate ambient RNA contamination using SoupX</p></td>
      </tr><tr><td>
          <p><code><a href="run_doubletFinder.html">run_doubletFinder()</a></code> </p>
        </td>
        <td><p>Run DoubletFinder</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_filtering_param.html">calculate_filtering_param()</a></code> </p>
        </td>
        <td><p>Estimate the filtering parameters (MT, nFeature, nCount)</p></td>
      </tr><tr><td>
          <p><code><a href="load_10x_gem.html">load_10x_gem()</a></code> </p>
        </td>
        <td><p>Load 10x GEM data</p></td>
      </tr><tr><td>
          <p><code><a href="run_qc_and_filtering.html">run_qc_and_filtering()</a></code> </p>
        </td>
        <td><p>Check sample quality and filter the count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="run_seurat_integration.html">run_seurat_integration()</a></code> </p>
        </td>
        <td><p>Samples integration using seurat</p></td>
      </tr><tr><td>
          <p><code><a href="run_seurat_integration_v5.html">run_seurat_integration_v5()</a></code> </p>
        </td>
        <td><p>Samples integration using seurat v5 (logNorm only)</p></td>
      </tr><tr><td>
          <p><code><a href="run_integration_with_Ref.html">run_integration_with_Ref()</a></code> </p>
        </td>
        <td><p>Samples reference based integration using seurat</p></td>
      </tr><tr><td>
          <p><code><a href="transfer_annotation.html">transfer_annotation()</a></code> </p>
        </td>
        <td><p>Function that takes the annotation from one dataset and it transfers it to another one.</p></td>
      </tr><tr><td>
          <p><code><a href="get_loupe_files.html">get_loupe_files()</a></code> </p>
        </td>
        <td><p>Dump umap coordinates and metadata to load in the loupe browser</p></td>
      </tr><tr><td>
          <p><code><a href="add_adt_assay.html">add_adt_assay()</a></code> </p>
        </td>
        <td><p>Add ADT information to a seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="add_vdj_assay.html">add_vdj_assay()</a></code> </p>
        </td>
        <td><p>Add VDJ information to a seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="CellCycleScoring_extra.html">CellCycleScoring_extra()</a></code> </p>
        </td>
        <td><p>Expansion of CellCycleScoring from the seurat package.</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_mito_ribo.html">calculate_mito_ribo()</a></code> </p>
        </td>
        <td><p>Calculate the percentage of mitochondrial and ribosomal genes</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_cell_complexity.html">calculate_cell_complexity()</a></code> </p>
        </td>
        <td><p>Calculate the complexity of the dataset</p></td>
      </tr><tr><td>
          <p><code><a href="compare_qc.html">compare_qc()</a></code> </p>
        </td>
        <td><p>Compare 2 seurat object (used to generate QC figures between pre-fitlered and filtered data)</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="annotation-functions">Annotation functions <a href="#annotation-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to annotate a dataset</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plot_cell_signature.html">plot_cell_signature()</a></code> </p>
        </td>
        <td><p>Plot gene signatures</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cells_markers.html">plot_cells_markers()</a></code> </p>
        </td>
        <td><p>Plot cell markers</p></td>
      </tr><tr><td>
          <p><code><a href="scType_pipeline.html">scType_pipeline()</a></code> </p>
        </td>
        <td><p>scType</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="archr-functions-scatac-seq-">ArchR functions (scATAC-Seq) <a href="#archr-functions-scatac-seq-" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Collection of functions to interact with ArchR projects</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="archR_subset_by_sample.html">archR_subset_by_sample()</a></code> </p>
        </td>
        <td><p>Subset an arcR projects based on sampleID value</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="vdj-functions-scvdj-seq-">VDJ functions (scVDJ-Seq) <a href="#vdj-functions-scvdj-seq-" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Collection of functions for TCR/BCR analysis</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="bcr_lineage_reconstruction.html">bcr_lineage_reconstruction()</a></code> </p>
        </td>
        <td><p>Build and visualise phylogenetic trees on AIRR formatted file</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="enrichment-analysis">Enrichment analysis <a href="#enrichment-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Useful functions for enrichment analysis (ORA + GSEA)</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="enrichR_analysis.html">enrichR_analysis()</a></code> </p>
        </td>
        <td><p>Run enrichR</p></td>
      </tr><tr><td>
          <p><code><a href="plot_enrichR_result.html">plot_enrichR_result()</a></code> </p>
        </td>
        <td><p>Generate the enrichment figure for enrichR</p></td>
      </tr><tr><td>
          <p><code><a href="fgsea_rank_genes.html">fgsea_rank_genes()</a></code> </p>
        </td>
        <td><p>Rank genes for FGSEA analysis</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualisation-functions">Visualisation functions <a href="#visualisation-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Mix of visualisation functions</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="VlnPlot_modified.html">VlnPlot_modified()</a></code> </p>
        </td>
        <td><p>Improved violin plot</p></td>
      </tr><tr><td>
          <p><code><a href="Clustered_DotPlot_nl.html">Clustered_DotPlot_nl()</a></code> </p>
        </td>
        <td><p>Improved Dotplot</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_cell_proportions.html">calculate_cell_proportions()</a></code> </p>
        </td>
        <td><p>Calculate cell proportions</p></td>
      </tr><tr><td>
          <p><code><a href="stemnet_plot.html">stemnet_plot()</a></code> </p>
        </td>
        <td><p>Plot a STEMNET object by Circular A Posteriori Projection</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="single-cell-utilities">Single-cell utilities <a href="#single-cell-utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Useful functions for the analysis of <strong>single-cell</strong> dataset</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calculate_genes_signature.html">calculate_genes_signature()</a></code> </p>
        </td>
        <td><p>Calcualte genes signature</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="interoperability-with-external-tools">Interoperability with external tools <a href="#interoperability-with-external-tools" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Useful functions that allow the interaction with external tools (e.g. scanpy)</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="export_seurat.html">export_seurat()</a></code> </p>
        </td>
        <td><p>Export a seurat object to file</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="bulk-utilities">Bulk utilities <a href="#bulk-utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Useful functions for the analysis of <strong>bulk</strong> dataset</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="PCA_genes.html">PCA_genes()</a></code> </p>
        </td>
        <td><p>Function that, given a DESeq2 object, runs PCA and it returns the contributions of the  
genes to each PCs.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="general-utilities">General utilities <a href="#general-utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Mix of functionalities</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="convert_ensemble_to_symbol.html">convert_ensemble_to_symbol()</a></code> </p>
        </td>
        <td><p>Convert the rownames of a matrix from ensemble to symbol</p></td>
      </tr><tr><td>
          <p><code><a href="mouse_human_gene_conversion.html">mouse_human_gene_conversion()</a></code> </p>
        </td>
        <td><p>Convert gene names from human to mouse</p></td>
      </tr><tr><td>
          <p><code><a href="SaveFigure.html">SaveFigure()</a></code> </p>
        </td>
        <td><p>Wrapper function to save figures to disk.</p></td>
      </tr><tr><td>
          <p><code><a href="Middle_Number.html">Middle_Number()</a></code> </p>
        </td>
        <td><p>Calcualte the middle value</p></td>
      </tr><tr><td>
          <p><code><a href="setup_sc_project.html">setup_sc_project()</a></code> </p>
        </td>
        <td><p>Setup project directory structure</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="datsets">Datsets <a href="#datsets" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>data used by the internal functions (mainly)</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ABIS_markers.html">ABIS_markers</a></code> </p>
        </td>
        <td><p>Gene markers of 29 immune cell types</p></td>
      </tr><tr><td>
          <p><code><a href="pbmc_markers.html">pbmc_markers</a></code> </p>
        </td>
        <td><p>PBMC markers</p></td>
      </tr><tr><td>
          <p><code><a href="pbmc_markers_annotated.html">pbmc_markers_annotated</a></code> </p>
        </td>
        <td><p>PBMC markers as data frame</p></td>
      </tr><tr><td>
          <p><code><a href="pbmc_1k.html">pbmc_1k</a></code> </p>
        </td>
        <td><p>PBMC 1K test dataset</p></td>
      </tr><tr><td>
          <p><code><a href="MouseHumanGenes.html">MouseHumanGenes</a></code> </p>
        </td>
        <td><p>Mouse-Human gene name mapping table</p></td>
      </tr><tr><td>
          <p><code><a href="ScTypeDB.html">ScTypeDB</a></code> </p>
        </td>
        <td><p>ScTypeDB</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="internal">Internal <a href="#internal" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Internal functions</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Parser.html">Parser</a></code> </p>
        </td>
        <td><p>Parsing the Garnett marker file</p></td>
      </tr><tr><td>
          <p><code><a href="validate_markers.html">validate_markers()</a></code> </p>
        </td>
        <td><p>Validate markers</p></td>
      </tr><tr><td>
          <p><code><a href="parse_input.html">parse_input()</a></code> </p>
        </td>
        <td><p>Parser</p></td>
      </tr><tr><td>
          <p><code><a href="parse_markers_lists.html">parse_markers_lists()</a></code> </p>
        </td>
        <td><p>Another parser</p></td>
      </tr><tr><td>
          <p><code><a href="markers_lookup.html">markers_lookup()</a></code> </p>
        </td>
        <td><p>Yes, another parser</p></td>
      </tr><tr><td>
          <p><code><a href="modify_scales.html">modify_scales()</a></code> </p>
        </td>
        <td><p>Modify scales of a Feature Plot.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Francesco Carbone.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

